File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/filter/healthelement/HealthElementByHcPartyTagCodeFilter.java
package org.taktik.icure.services.external.rest.v1.dto.filter.healthelement;

import com.google.common.base.Objects;
import org.taktik.icure.entities.HealthElement;
import org.taktik.icure.services.external.rest.handlers.JsonPolymorphismRoot;
import org.taktik.icure.services.external.rest.v1.dto.filter.Filter;

@JsonPolymorphismRoot(Filter.class)
public class HealthElementByHcPartyTagCodeFilter extends Filter<HealthElement> implements org.taktik.icure.dto.filter.healthelement.HealthElementByHcPartyTagCodeFilter {
	String healthCarePartyId;
	String codeType;
	String codeNumber;
	String tagType;
	String tagCode;
	Integer status;

	public HealthElementByHcPartyTagCodeFilter() {
	}

	public HealthElementByHcPartyTagCodeFilter(String healthCarePartyId, String codeType, String codeNumber, String tagType, String tagCode) {
		this.healthCarePartyId = healthCarePartyId;
		this.codeType = codeType;
		this.codeNumber = codeNumber;
		this.tagType = tagType;
		this.tagCode = tagCode;
	}

	@Override
	public String getHealthCarePartyId() {
		return healthCarePartyId;
	}

	@Override
	public String getCodeType() {
		return codeType;
	}

	@Override
	public String getCodeNumber() {
		return this.codeNumber;
	}

	@Override
	public String getTagType() {
		return this.tagType;
	}

	@Override
	public String getTagCode() {
		return this.tagCode;
	}

	@Override
	public Integer getStatus() {
		return this.status;
	}

	public void setHealthCarePartyId(String healthCarePartyId) {
		this.healthCarePartyId = healthCarePartyId;
	}

	public void setCodeType(String codeType) {
		this.codeType = codeType;
	}

	public void setCodeNumber(String codeNumber) {
		this.codeNumber = codeNumber;
	}

	public void setTagType(String tagType) {
		this.tagType = tagType;
	}

	public void setTagCode(String tagCode) {
		this.tagCode = tagCode;
	}

	public void setStatus(Integer status) {
		this.status = status;
	}

	@Override
	public int hashCode() {
		return Objects.hashCode(healthCarePartyId, codeType, codeNumber, tagType, tagCode, status);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null || getClass() != obj.getClass()) {
			return false;
		}
		final HealthElementByHcPartyTagCodeFilter other = (HealthElementByHcPartyTagCodeFilter) obj;
		return Objects.equal(this.healthCarePartyId, other.healthCarePartyId) && Objects.equal(this.codeType, other.codeType) && Objects.equal(this.codeNumber, other.codeNumber)
				&& Objects.equal(this.tagType, other.tagType) && Objects.equal(this.tagCode, other.tagCode) && Objects.equal(this.status, other.status);
	}

	@Override
	public boolean matches(HealthElement item) {
		return (healthCarePartyId == null || item.getDelegations().keySet().contains(healthCarePartyId))
				&& (codeType == null || (item.getCodes().stream().filter(code -> codeType.equals(code.getType()) && codeNumber.equals(code.getCode())).findAny().isPresent())
				&& (tagType == null || item.getTags().stream().filter(t -> tagType.equals(t.getType()) && (tagCode == null || tagCode.equals(t.getCode()))).findAny().isPresent())
				&& (status == null || item.getStatus() == status));
	}
}

File: src/main/kotlin/org/taktik/icure/services/external/rest/v1/dto/filter/healthelement/HealthElementByHcPartyTagCodeFilter.kt
package org.taktik.icure.services.external.rest.v1.dto.filter.healthelement

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.github.pozo.KotlinBuilder
import org.taktik.icure.entities.HealthElement
import org.taktik.icure.handlers.JsonPolymorphismRoot
import org.taktik.icure.services.external.rest.v1.dto.filter.AbstractFilterDto

@JsonPolymorphismRoot(AbstractFilterDto::class)
@JsonDeserialize(using = JsonDeserializer.None::class)
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
data class HealthElementByHcPartyTagCodeFilter(
        override val desc:String? = null,
        override val healthCarePartyId: String? = null,
        override val codeType: String? = null,
        override val codeNumber: String? = null,
        override val tagType: String? = null,
        override val tagCode: String? = null,
        override val status: Int? = null
        ) : AbstractFilterDto<HealthElement>, org.taktik.icure.domain.filter.healthelement.HealthElementByHcPartyTagCodeFilter
